components/tidb_query/src/codec/mysql/time/datetime.rs

```diff
@@ -0,0 +1,1243 @@
+use std::convert::TryFrom;
+use std::sync::Arc;
+
+use crate::codec;
+use crate::codec::mysql::{self, Tz};
+use crate::codec::TEN_POW;
+use crate::codec::{Error, Result};
+use crate::expr::{EvalConfig, EvalContext, Flag, SqlMode};
+
+use tidb_query_datatype::FieldTypeTp;
+
+use bitfield::bitfield;
+use chrono::prelude::*;
+
+const MIN_TIMESTAMP: i64 = 0;
+const MAX_TIMESTAMP: i64 = (1 << 31) - 1;
+const MICRO_WITDH: usize = 6;
+
+fn bytes_to_u32(input: &[u8]) -> Result<u32> {
+    unsafe {
+        std::str::from_utf8_unchecked(input)
+            .parse()
+            .map_err(|_| Error::truncated())
+    }
+}
+
+fn is_leap_year(year: u32) -> bool {
+    year & 3 == 0 && (year % 100 != 0 || year % 400 == 0)
+}
+
+fn last_day_of_month(year: u32, month: u32) -> u32 {
+    match month {
+        4 | 6 | 9 | 11 => 30,
+        2 => is_leap_year(year) as u32 + 28,
+        _ => 31,
+    }
+}
+
+bitfield! {
+    #[derive(Clone, Copy, Default)]
+    pub struct Time(u64);
+
+    u32;
+    #[inline]
+    get_year, set_year: 63, 50;
+    #[inline]
+    get_month, set_month: 49, 46;
+    #[inline]
+    get_day, set_day: 45, 41;
+    #[inline]
+    get_hour, set_hour: 40, 36;
+    #[inline]
+    get_minute, set_minute: 35, 30;
+    #[inline]
+    get_second,set_second: 29, 24;
+    #[inline]
+    get_micro, set_micro: 23, 4;
+
+    // `fsp_tt` format:
+    // | fsp: 3 bits | type: 1 bit |
+    // When `fsp` is valid (in range [0, 6]):
+    // 1. `type` bit 0 represent `DateTime`
+    // 2. `type` bit 1 represent `TimeStamp`
+    //
+    // Since `Date` does not require `fsp`, we could use `fsp` == 0b111 to represent it.
+    #[inline]
+    u8, get_fsp_tt, set_fsp_tt: 3, 0;
+}
+
+#[derive(Debug)]
+struct DateTimeConfig {
+    strict_mode: bool,
+    no_zero_in_date: bool,
+    no_zero_date: bool,
+    allow_invalid_date: bool,
+    ignore_truncate: bool,
+    time_zone: Option<Tz>,
+}
+
+impl DateTimeConfig {
+    pub fn from_ctx(ctx: &EvalContext) -> DateTimeConfig {
+        let sql_mode = ctx.cfg.sql_mode;
+        let flags = ctx.cfg.flag;
+        DateTimeConfig {
+            strict_mode: sql_mode.contains(SqlMode::STRICT_ALL_TABLES),
+            no_zero_in_date: sql_mode.contains(SqlMode::NO_ZERO_IN_DATE),
+            no_zero_date: sql_mode.contains(SqlMode::NO_ZERO_DATE),
+            allow_invalid_date: sql_mode.contains(SqlMode::INVALID_DATES),
+            ignore_truncate: flags.contains(Flag::IGNORE_TRUNCATE),
+            time_zone: Some(ctx.cfg.tz.clone()),
+        }
+    }
+}
+
+impl From<DateTimeConfig> for EvalContext {
+    fn from(config: DateTimeConfig) -> EvalContext {
+        let mut eval_config = EvalConfig::new();
+        let mut sql_mode = SqlMode::empty();
+        let mut flags = Flag::empty();
+        eval_config.tz = config.time_zone.unwrap_or_else(Tz::utc);
+
+        if config.strict_mode {
+            sql_mode |= SqlMode::STRICT_ALL_TABLES;
```

Comment:
I think we can use `config.strict_mode as i64 * SqlMode::STRICT_ALL_TABLES` here. To avoid jmp or condition move